src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,276 @@
+                                        uint8_t msgType, System::PacketBuffer * payload);
+    MessageReceiveFunct OnMessageReceived;
+
+    /**
+     * This function is the protocol callback to invoke when the timeout for the receipt
+     * of a response message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*ResponseTimeoutFunct)(ExchangeContext * ec);
+    ResponseTimeoutFunct OnResponseTimeout;
+
+    /**
+     * This function is the protocol callback to invoke when the timeout for the retransmission
+     * of a previously sent message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*RetransmissionTimeoutFunct)(ExchangeContext * ec);
+    RetransmissionTimeoutFunct OnRetransmissionTimeout;
+
+    /**
+     * Type of key error message handling function.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    keyErr        The CHIP_ERROR type that was reported in the key error message.
+     *
+     */
+    typedef void (*KeyErrorFunct)(ExchangeContext * ec, CHIP_ERROR keyErr);
+
+    /**
+     * This function is the protocol callback to invoke when key error message has been received
+     * from the peer.
+     */
+    KeyErrorFunct OnKeyError;
+
+    void CancelRetrans();
+
+    /*
+     * in order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
+    void Close();
+    void Abort();
+    void Release();
+    CHIP_ERROR StartTimerT();
+
+private:
+    System::PacketBuffer * msg; // If we are re-transmitting, then this is the pointer to the message being retransmitted
+
+    // Trickle-controlled retransmissions:
+    uint32_t backoff;                        // backoff for sampling the numner of messages
+    uint8_t msgsReceived;                    // number of messages heard during the backoff period
+    uint8_t rebroadcastThreshold;            // re-broadcast threshold
+    BitFlags<uint16_t, ExFlagValues> mFlags; // Internal state flags
+
+    CHIP_ERROR ResendMessage();
+    bool MatchExchange(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader);
+    static void CancelRetransmissionTimer(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerTau(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerT(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    CHIP_ERROR StartResponseTimer();
+    void CancelResponseTimer();
+    static void HandleResponseTimeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    uint32_t mPendingPeerAckId;
```

Comment:
What is this AckId tracking? Is it for CRMP use?

Reply:
Yes, this is for CRMP, remove it for now.